README.md
# Google Photos API client for Go
[![Go Documentation](https://img.shields.io/badge/go-documentation-blue.svg?style=flat-square)](https://godoc.org/github.com/gphotosuploader/google-photos-api-client-go/lib-gphotos)
[![Build Status](https://cloud.drone.io/api/badges/gphotosuploader/google-photos-api-client-go/status.svg)](https://cloud.drone.io/gphotosuploader/google-photos-api-client-go)
[![Go Report Card](https://goreportcard.com/badge/github.com/gphotosuploader/google-photos-api-client-go)](https://goreportcard.com/report/github.com/gphotosuploader/google-photos-api-client-go)
[![codebeat badge](https://codebeat.co/badges/c0ab08dd-11b3-406e-bbcc-b9d4a90aedf6)](https://codebeat.co/projects/github-com-gphotosuploader-google-photos-api-client-go-master)
[![codecov](https://codecov.io/gh/gphotosuploader/google-photos-api-client-go/branch/master/graph/badge.svg)](https://codecov.io/gh/gphotosuploader/google-photos-api-client-go)
[![GitHub release](https://img.shields.io/github/release/gphotosuploader/google-photos-api-client-go.svg)](https://github.com/gphotosuploader/google-photos-api-client-go/releases/latest)
[![GitHub](https://img.shields.io/github/license/gphotosuploader/google-photos-api-client-go.svg)](LICENSE)

This package provides a client for using the [Google Photos API](https://godoc.org/google.golang.org/api). Uses the [original Google Photos package](https://github.com/denysvitali/go-googlephotos), that was provided by Google, and [removed](https://code-review.googlesource.com/c/google-api-go-client/+/39951) some time ago. On top of the old package has been extended some other features like uploads, including resumable uploads.


> This project will maintain compatibility with the last two Go major versions published. Currently Go 1.12 and Go 1.13. 
>
## Quick start

Construct a new Google Photos client, then use the various services on the client to access different parts of the Google Photos API. For example:

```go
	import "github.com/gphotosuploader/google-photos-api-client-go/pkg"

    // httpClient is an authenticated http.Client. See Authentication below.
	client := gphotos.NewClient(httpClient)
    // get or create a Photos Album with the specified name.
	album, err := GetOrCreateAlbumByName("my-new-album")
	// upload an specified file to an existent Photos Album.
    _, err := client.AddMediaItem(ctx, path, albumID)
```

> NOTE: Using the [context package](https://godoc.org/context), one can easily pass cancellation signals and deadlines to various services of the client for handling a request. In case there is no context available, then `context.Background()` can be used as a starting point.

## Authentication
The gphotos library **does not directly handle authentication**. Instead, when creating a new client, pass an `http.Client` that can handle authentication for you. The easiest and recommended way to do this is using the `golang.org/x/oauth2` library, but you can always use any other library that provides an `http.Client`.

Access to the API requires OAuth client credentials from a Google developers project. This project must have the Library API enabled as described [here](https://developers.google.com/photos/library/guides/get-started).

```go
	import (
        "golang.org/x/oauth2"
        "github.com/gphotosuploader/google-photos-api-client-go/pkg"
    )
	func main() {
		ctx := context.Background()
		oc := oauth2Config := oauth2.Config{
			ClientID:     "... your application Client ID ...",
			ClientSecret: "... your application Client Secret ...",
			Endpoint:     photos.Endpoint,
			Scopes:       photos.Scopes,
		}
		tc := oc.Client(ctx, "... your user Oauth Token ...")
		client := gphotos.NewClient(tc)
		// look for a Google Photos Album by name
		album, _, err := client.AlbumByName(ctx, "my-album")
	}
```

Note that when using an authenticated Client, all calls made by the client will include the specified OAuth token. Therefore, authenticated clients should almost never be shared between different users. See the oauth2 docs for complete instructions on using that library.

## Limitations
### Rate Limiting
Google Photos imposes a rate limit on all API clients. The quota limit for requests to the Library API is 10,000 requests per project per day. The quota limit for requests to access media bytes (by loading a photo or video from a base URL) is 75,000 requests per project per day.

### Photo storage and quality
All media items uploaded to Google Photos using the API [are stored in full resolution]((https://support.google.com/photos/answer/6220791)) at original quality. **They count toward the user’s storage**.


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/) and this project adheres to [Semantic Versioning](https://semver.org/).

## 1.1.5
### Changed
- Update required `googlemirror` package version to v0.3.5.

## 1.1.4
> This version was not working due to `photoslibrary` dependency. PLEASE UPDATE TO v1.1.5 ASAP.
### Changed
- Update required `googlemirror` package version to v0.3.4.

## 1.1.3
### Added
- [CONTRIBUTING](CONTRIBUTING.md) guide line has been added.
### Changed
- [README](README.md) has been updated fixing some typos.
- Module use an interface Logger to log activity. This allow to implement different logging systems. See [internal/log/logger.go](v2/internal/log/logger.go).
### Deprecated
- Once Go 1.13 has been published, previous Go 1.11 support is deprecated. This project will maintain compatibility with the last two major versions published.
### Fix
- Fix uploaded photos without a correct file name. ([#32][i32])
### Removed
- Remove progress information when uploading, if you want to have progress bars when upload a file, you should inject a reader to the Uploader().

[i32]: https://github.com/gphotosuploader/google-photos-api-client-go/issues/32

## 1.1.2
### Changed
- Update [golangci](https://github.com/golangci/golangci-lint) linter to version 1.20.0.
### Fixed
- Fix fatal error introduced in the last version. (#28)

## 1.1.1
### Fixed
- Fix race condition on `GetOrCreateAlbumByName()`. Google Photos API allow you to create several albums with the same name. (#26)

## 1.1.0
### Added
- New `NewClientWithResumableUploads()` function to create a Google Photos client with uploads that can be resumed.
- New `AddMediaItem()` method to upload contents and make it available in Google Photos.
- New `uploader` internal package implementing uploads to Google Photos.
### Changed
- Refactor how Google response is checked. It's following Google Photos best practices. (#10)
- Updated README documentation and added examples.
- Refactor how retries are handled by the code. See `retryableMediaItemBatchCreateDo()` method.
### Deprecated
- `NewClient()` function has been deprecated. Use `NewClientWithResumableUploads()` instead.
- `UploadFile()` and `UploadFileResumable()` methods has been deprecated. Use `AddMediaItem()` instead.
- `noserver-gphotos` package has been deprecated.

## 1.0.7
### Added
- A new Logger has been added to be shared across the whole package. See `logger.go`.
- Tests for almost all the code, except `uploads.go` that needs a lot of refactor.
- Package's documentation and examples of use.

### Changed
- `xerrors` is used instead of `errors`. Preparing code for Go 1.13 release.

### Deprecated
- `Token()` method has been deprecated. Current package implementation doesn't need to have OAuth token.

## 1.0.6
### Fixed
- Fix progress logging. (#19)

## 1.0.5
### Added
- Code quality reporting using [Codebeat](https://codebeat.co/projects/github-com-gphotosuploader-google-photos-api-client-go-master).
- Code coverage reporting using [codecov](https://codecov.io/gh/gphotosuploader/google-photos-api-client-go).

### Fixed
- `ReadProgressReporter` was giving panic at some circumstances. There was a problem casting `int64` to `int`. (#17)
- Fix progress calculations for files big sized. 

### Changed
- CI platform is now [drone.io](https://drone.io).

## 1.0.4
### Fixed
- Fix `AlbumByName` to check against all Google Photos album list (#12).

## 1.0.3
### Added
- Add resumable file uploads. You can use new `UploadFileResumable` method to upload files that can be resumed. See [documentation](https://godoc.org/github.com/gphotosuploader/google-photos-api-client-go/lib-gphotos) for more details.

## 1.0.2
### Added
- Add makefile for easy test and linting
- Add CI using travis-ci.com
- Add travis and goodocs badges to README

## 1.0.1
### Added
- Add semantic versioning to this package
- Add Go module support
- Add MIT license

### Changed
- Clean up of useless files / directories
- Update package documentation

### Fixed
- Fix issue #8 on parent repository [here](https://github.com/nmrshll/google-photos-api-client-go/issues/8)

### Removed
- Removed `Makefile` support

## 1.0.0
Initial release after clone it from [original repository](https://github.com/nmrshll/google-photos-api-client-go). Latest commit was [3dac07f](https://github.com/nmrshll/google-photos-api-client-go/commit/3dac07f1b07f249ac4a9805b9d60afe0f68c34b2)


v2/doc.go
/*
Package gphotos provides a client for using the Google Photos API.
Wraps the Google Photos package provided by Google, and now removed:
https://github.com/denysvitali/go-googlephotos.

Usage:
	import "github.com/gphotosuploader/google-photos-api-client-go/pkg"

Construct a new Google Photos client, then use the various services on the client to
access different parts of the Google Photos API. For example:

    // httpClient is an authenticated http.Client. See Authentication below.
	client := gphotos.NewClient(httpClient)
    // get or create a Photos Album with the specified name.
	album, err := GetOrCreateAlbumByName("my-new-album")
	// upload an specified file to an existent Photos Album.
    _, err := client.AddMediaItem(ctx, path, albumID)


NOTE: Using the https://godoc.org/context package, one can easily
pass cancellation signals and deadlines to various services of the client for
handling a request. In case there is no context available, then context.Background()
can be used as a starting point.

Authentication
The gphotos library does not directly handle authentication. Instead, when
creating a new client, pass an http.Client that can handle authentication for
you. The easiest and recommended way to do this is using the golang.org/x/oauth2
library, but you can always use any other library that provides an http.Client.
Access to the API requires OAuth client credentials from a Google developers
project. This project must have the Library API enabled as described in
https://developers.google.com/photos/library/guides/get-started.

	import "golang.org/x/oauth2"
	func main() {
		ctx := context.Background()
		oc := oauth2Config := oauth2.Config{
			ClientID:     "... your application Client ID ...",
			ClientSecret: "... your application Client Secret ...",
			Endpoint:     photos.Endpoint,
			Scopes:       photos.Scopes,
		}
		tc := oc.Client(ctx, "... your user Oauth Token ...")
		client := gphotos.NewClient(tc)
		// look for a Google Photos Album by name
		album, _, err := client.AlbumByName(ctx, "my-album")
	}

Note that when using an authenticated Client, all calls made by the client will
include the specified OAuth token. Therefore, authenticated clients should
almost never be shared between different users.
See the oauth2 docs for complete instructions on using that library.

Rate Limiting
Google Photos imposes a rate limit on all API clients. The quota limit for
requests to the Library API is 10,000 requests per project per day. The quota
limit for requests to access media bytes (by loading a photo or video from a base
URL) is 75,000 requests per project per day.

Photo storage and quality
All media items uploaded to Google Photos using the API are stored in full
resolution at original quality (https://support.google.com/photos/answer/6220791).
They count toward the user’s storage.
*/
package google_photos_api_client_go
